### PR TITLE
dyno: Yet more fixes for parser and converter

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
dyno: Yet more fixes for parser and converter (#19711)

This PR is a collection of many different small patches which aim
to reduce test failures when running under `--dyno`, `LLVM=system`.
It modifies both parsers and the converter.

This PR incorporates prior commits by @mppf which adjust the old
scanner to be consistent with dyno error messages.

It also includes his minor adjustment to the spec for the
"hexadecimal digit" rule.

[Summary: #Successes = 13532 | #Failures = 37 | #Futures = 0]
(with `LLVM=system`)

TESTING

- [x] `ALL` on `linux64` w/ `COMM=none`
- [x] `ALL` on `linux64` w/ `COMM=gasnet`

Reviewed by @arezaii. Thanks!

Signed-off-by: David Longnecker <dlongnecke-cray@users.noreply.github.com>

### CODE REVIEW

**Comment by dlongnecke-cray** [comment]
I expect the CI checks to fail until I rebuild the parser(s).

File: compiler/dyno/test/parsing/testParseStringBytesLiterals.cpp [hunk 1]
```diff
@@ -150,6 +150,7 @@ static void testSingleLiteral(Parser* parser,
 static void testBadLiteral(Parser* parser,
                            const char* testname,
                            const char* str) {
+  if (!strcmp(testname, "test32.chpl")) debuggerBreakHere();
```
Comment (dlongnecke-cray) [inline compiler/dyno/test/parsing/testParseStringBytesLiterals.cpp:153]: Oof, forgot to remove this line.

**Comment by dlongnecke-cray** [comment]
A few test failures on `--no-dyno` to attend to:

```
  2 [Error matching compiler output for classes/initializers/postInit/invalidSuperPostinit]
  3 [Error matching compiler output for types/string/ferguson/literal-octal]
  4 [Error matching compiler output for types/string/ferguson/literal-unicode]
```

File: compiler/dyno/test/uast/testStringify.cpp [hunk 2]
```diff
@@ -385,22 +385,31 @@ static void test2(Parser* parser) {
...
+  TEST_USER_STRING("proc procRefC(ref arg: borrowed C?) { }",
+                   "procRefC(ref arg: borrowed C?)")
   TEST_USER_STRING("operator =(ref r3:R, r4:R){}", " = (ref r3: R, r4: R)")
   TEST_USER_STRING("proc foo(x...) {}", "foo(x ...)")
   TEST_USER_STRING("proc bar(type x...) {}", "bar(type x ...)")
   TEST_USER_STRING("proc bar(type x...?k) {}", "bar(type x ...?k)")
-  TEST_USER_STRING("proc foo(x: borrowed C(t=?tt, r=?rr), y: borrowed C(tt, rr)) {}", "foo(x: borrowed C(t = ?tt, r = ?rr), y: borrowed C(tt, rr))")
+  TEST_USER_STRING("proc foo(x: borrowed C(t=?tt, r=?rr), y: borrowed C(tt, rr)) {}",
+                   "foo(x: borrowed C(t = ?tt, r = ?rr), y: borrowed C(tt, rr))")
 }
```
Comment (arezaii) [inline compiler/dyno/test/uast/testStringify.cpp:413]: can we add a test to capture the fix you made for (int(32)).foo()?

should look like:
```
TEST_USER_STRING("proc (int(32)).foo() {/n}", "(int(32)).foo()")
```

Comment (dlongnecke-cray) [reply compiler/dyno/test/uast/testStringify.cpp:414]: Added, thanks!

**Review by arezaii — APPROVED** [review]
A major effort indeed. It looks good to me, but we may want a second pair of eyes, just due to the number of changes

File: compiler/dyno/lib/parsing/ParserContextImpl.h [hunk 3]
```diff
@@ -800,12 +810,22 @@ CommentsAndStmt ParserContext::buildFunctionDecl(YYLTYPE location,
...
-    auto f = Function::build(builder, this->convertLocation(location),
+    // Own the recorded identifier to clean it up, but grab its location.
+    owned<Identifier> identName = toOwned(fp.name);
+    assert(identName.get());
+    auto identNameLoc = builder->getLocation(identName.get());
+    assert(!identNameLoc.isEmpty());
+
+    // TODO: Right now this location is the start of the function, and
+    // it seems more natural for the location to be the symbol.
+    (void) location;
```
Comment (mppf) [inline compiler/dyno/lib/parsing/ParserContextImpl.h:821]: Why is this statement here?

Comment (dlongnecke-cray) [reply compiler/dyno/lib/parsing/ParserContextImpl.h:821]: I have this habit of voiding things that aren't really used to call them out for later removal. I removed it for you :)

**Pull request merged**